File: frontend/src/services/citation.service.ts
import type { Citation } from "@/types/chat.types"

export interface ConfidenceBadgeInfo {
  scorePercent: number
  level: "High" | "Medium" | "Fair" | "Low"
  colorClass: string
  bgClass: string
  borderClass: string
}

export class CitationService {
  /**
   * Calculates confidence badge info based on float confidence (0.0 to 1.0) or percentage (0 to 100)
   */
  static getConfidenceBadge(confidence?: number, level?: string): ConfidenceBadgeInfo {
    const rawScore = confidence ?? 0.95
    const scorePercent = Math.round(rawScore <= 1.0 ? rawScore * 100 : rawScore)

    let calcLevel: "High" | "Medium" | "Fair" | "Low" = (level as any) || "High"
    if (!level) {
      if (scorePercent >= 95) calcLevel = "High"
      else if (scorePercent >= 80) calcLevel = "Medium"
      else if (scorePercent >= 60) calcLevel = "Fair"
      else calcLevel = "Low"
    }

    let colorClass = "text-emerald-600 dark:text-emerald-400"
    let bgClass = "bg-emerald-500/10"
    let borderClass = "border-emerald-500/30"

    if (calcLevel === "Medium") {
      colorClass = "text-blue-600 dark:text-blue-400"
      bgClass = "bg-blue-500/10"
      borderClass = "border-blue-500/30"
    } else if (calcLevel === "Fair") {
      colorClass = "text-amber-600 dark:text-amber-400"
      bgClass = "bg-amber-500/10"
      borderClass = "border-amber-500/30"
    } else if (calcLevel === "Low") {
      colorClass = "text-rose-600 dark:text-rose-400"
      bgClass = "bg-rose-500/10"
      borderClass = "border-rose-500/30"
    }

    return {
      scorePercent,
      level: calcLevel,
      colorClass,
      bgClass,
      borderClass,
    }
  }

  /**
   * Constructs deep link URL into Document Viewer
   */
  static buildDeepLinkUrl(citation: Citation): string {
    const docId = citation.document_id || "preview"
    const page = citation.page_number || 1
    const chunk = citation.chunk_index ?? 0
    const highlightParam = citation.excerpt ? `&highlight=${encodeURIComponent(citation.excerpt.slice(0, 100))}` : ""

    return `/documents/${docId}?page=${page}&chunk=${chunk}${highlightParam}`
  }

  /**
   * Builds Workspace > Folder > Document > Heading > Page > Chunk breadcrumbs
   */
  static getBreadcrumbs(citation: Citation): string[] {
    const crumbs: string[] = []
    crumbs.push(citation.workspace_name || "Workspace")
    crumbs.push(citation.folder_name || "Folder")
    crumbs.push(citation.document_name)
    if (citation.heading) crumbs.push(citation.heading)
    if (citation.page_number) crumbs.push(`Trang ${citation.page_number}`)
    if (citation.chunk_index !== undefined) crumbs.push(`Chunk #${citation.chunk_index}`)
    return crumbs
  }
}

File: backend/app/services/citation_service.py
import re
from typing import Any, Dict, List, Tuple
from uuid import UUID
from app.schemas.citation import CitationSchema


class CitationService:
    @staticmethod
    def calculate_confidence(score: float) -> Tuple[float, str]:
        """
        Converts similarity score into a normalized float (0.0 to 1.0)
        and returns confidence label:
        High (95-100), Medium (80-94), Fair (60-79), Low (<60)
        """
        norm_score = max(0.0, min(1.0, score if score <= 1.0 else score / 100.0))
        percentage = norm_score * 100.0

        if percentage >= 95.0:
            level = "High"
        elif percentage >= 80.0:
            level = "Medium"
        elif percentage >= 60.0:
            level = "Fair"
        else:
            level = "Low"

        return norm_score, level

    @staticmethod
    def extract_excerpt(text: str, max_words: int = 40) -> str:
        """Extracts a clean, representative excerpt from chunk content."""
        if not text:
            return "Trích đoạn từ tài liệu chỉ mục RAG..."
        clean = re.sub(r"\s+", " ", text).strip()
        words = clean.split(" ")
        if len(words) <= max_words:
            return clean
        return " ".join(words[:max_words]) + "..."

    @staticmethod
    def enrich_citation(
        doc_name: str,
        source_type: str = "pdf",
        page_number: int | None = 1,
        heading: str | None = None,
        workspace_name: str | None = None,
        folder_name: str | None = None,
        document_id: Any | None = None,
        chunk_id: str | None = None,
        chunk_index: int | None = 0,
        score: float = 0.95,
        excerpt: str | None = None,
        url: str | None = None,
        line_start: int | None = None,
        line_end: int | None = None,
    ) -> CitationSchema:
        """Creates a fully enriched CitationSchema object."""
        confidence_val, level = CitationService.calculate_confidence(score)
        clean_excerpt = excerpt or "Đoạn văn bản trích xuất từ chỉ mục RAG..."

        return CitationSchema(
            document_id=str(document_id) if document_id else None,
            document_name=doc_name or "Tài liệu",
            workspace_name=workspace_name or "General",
            folder_name=folder_name or "Root",
            page_number=page_number or 1,
            heading=heading or f"Mục {page_number or 1}",
            chunk_id=chunk_id or f"chunk-{chunk_index or 0}",
            chunk_index=chunk_index or 0,
            confidence=confidence_val,
            confidence_level=level,
            source_type=source_type or "pdf",
            excerpt=CitationService.extract_excerpt(clean_excerpt),
            url=url,
            line_start=line_start or 1,
            line_end=line_end or 10,
        )

    @staticmethod
    def format_breadcrumb(citation: CitationSchema) -> List[str]:
        """Generates Workspace > Folder > Document > Heading > Page > Chunk hierarchy."""
        crumbs = []
        if citation.workspace_name:
            crumbs.append(citation.workspace_name)
        if citation.folder_name:
            crumbs.append(citation.folder_name)
        crumbs.append(citation.document_name)
        if citation.heading:
            crumbs.append(citation.heading)
        if citation.page_number:
            crumbs.append(f"Trang {citation.page_number}")
        if citation.chunk_index is not None:
            crumbs.append(f"Chunk #{citation.chunk_index}")
        return crumbs
